src/hotspot/share/gc/g1/g1YoungGCPostEvacuateTasks.cpp

```diff
@@ -97,24 +97,28 @@ class G1PostEvacuateCollectionSetCleanupTask1::SampleCollectionSetCandidatesTask
   }
 };
 
-class G1PostEvacuateCollectionSetCleanupTask1::RemoveSelfForwardPtrsTask : public G1AbstractSubTask {
-  G1ParRemoveSelfForwardPtrsTask _task;
+class G1PostEvacuateCollectionSetCleanupTask1::RestoreRetainedRegionsTask : public G1AbstractSubTask {
+  G1RemoveSelfForwardsInChunksTask _task;
   G1EvacFailureRegions* _evac_failure_regions;
 
 public:
-  RemoveSelfForwardPtrsTask(G1EvacFailureRegions* evac_failure_regions) :
+  RestoreRetainedRegionsTask(G1EvacFailureRegions* evac_failure_regions) :
     G1AbstractSubTask(G1GCPhaseTimes::RestoreRetainedRegions),
     _task(evac_failure_regions),
     _evac_failure_regions(evac_failure_regions) { }
 
   double worker_cost() const override {
     assert(_evac_failure_regions->evacuation_failed(), "Should not call this if not executed");
-    return _evac_failure_regions->num_regions_failed_evacuation();
+    return _evac_failure_regions->num_regions_failed_evacuation() * G1PerRetainedRegionThreads;
```

Comment:
It feels a bit wrong to have a flag for this, or at least I'm not sure if the number of thread only should be connected to the number of regions only. Instead maybe we should look at the to total size for those regions. Now with the chunking there will be a lot more work items than regions so we should be able to look at size which to me looks like a better metric.

Reply:
Something like `G1CollectedHeap::get_chunks_per_region() / MAGIC_FACTOR` I guess?

Reply:
Yes, that would be better and times the number of regions.

Reply:
Done in the latest change. However I kept a diagnostic flag to allow tuning of that. I'll do some more testing to see whether it makes a difference.